server/testkit/tests/identity_verification_tests.rs

```diff
@@ -0,0 +1,364 @@
+async fn test_start_response_wait_for_code_or_provide_code(rsclient: KanidmClient) {
+    setup_server(&rsclient).await;
+    let user_a_uuid = create_user(&rsclient, USER_A_NAME).await;
+    let user_b_uuid = create_user(&rsclient, USER_B_NAME).await;
+    login_with_user(&rsclient, USER_A_NAME).await;
+    let response = rsclient
+        .idm_person_identify_user(USER_B_NAME, IdentifyUserRequest::Start)
+        .await;
+
+    assert!(response.is_ok());
+    // the person with the lowest uuid should get to input the other person's code first;
+    dbg!(user_a_uuid.clone(), user_b_uuid.clone());
+
+    if user_a_uuid < user_b_uuid {
+        assert_eq!(response.unwrap(), IdentifyUserResponse::WaitForCode);
+    } else {
+        assert!(matches!(
+            response.unwrap(),
+            IdentifyUserResponse::ProvideCode { .. }
+        ))
+    }
+}
+
+#[kanidmd_testkit::test]
+async fn test_provide_code_response_code_failure_or_provide_code(rsclient: KanidmClient) {
+    setup_server(&rsclient).await;
+    let user_a_uuid = create_user(&rsclient, USER_A_NAME).await;
+    let user_b_uuid = create_user(&rsclient, USER_B_NAME).await;
+    login_with_user(&rsclient, USER_A_NAME).await;
+    let response = rsclient
+        .idm_person_identify_user(
+            USER_B_NAME,
+            IdentifyUserRequest::SubmitCode { other_totp: 123456 },
+        )
+        .await;
+    //if A is the first then either the code is correct and therefore we get a ProvideCode or it's wrong
+    // and we get a CodeFailure
+    if user_a_uuid < user_b_uuid {
+        assert!(matches!(
+            response.unwrap(),
+            IdentifyUserResponse::ProvideCode { .. } | IdentifyUserResponse::CodeFailure
+        ));
+    } else {
+        assert!(matches!(
+            response.unwrap(),
+            IdentifyUserResponse::Success | IdentifyUserResponse::CodeFailure
+        ));
+    }
+}
+
+// here we actually test the full idm flow by duplicating the server
+#[kanidmd_testkit::test]
+async fn test_full_identification_flow(rsclient: KanidmClient) {
+    setup_server(&rsclient).await;
+    let user_a_uuid = create_user(&rsclient, USER_A_NAME).await;
+    let user_b_uuid = create_user(&rsclient, USER_B_NAME).await;
+    //user A session
+    let valid_user_a_client = rsclient;
+    login_with_user(&valid_user_a_client, USER_A_NAME).await;
+    //user B session
+    let valid_user_b_client = valid_user_a_client.new_session().unwrap();
+    login_with_user(&valid_user_b_client, USER_B_NAME).await;
+
+    let user_a_req_1 = valid_user_a_client
+        .idm_person_identify_user(USER_B_NAME, IdentifyUserRequest::Start)
+        .await
+        .unwrap();
+
+    let user_b_req_1 = valid_user_b_client
+        .idm_person_identify_user(USER_A_NAME, IdentifyUserRequest::Start)
+        .await
+        .unwrap();
+
+    // now we have to consider the two separate cases: first we address the case a has the lowest uuid
+    if user_a_uuid < user_b_uuid {
```

Comment:
I wonder if there is a way to do this neater, like having a tuple that swaps the clients like:

```
let (client_left, client right) = if user_a_uuid < user_b_uuid {
    client_a, client_b
} else {
    client_b, client_a
};
```

Reply:
niiice way better than how I was doing it before